src/dmd/root/longdouble.d

```diff
@@ -6,29 +6,92 @@
  * https://github.com/dlang/dmd/blob/master/src/root/longdouble.d
  */
 
-// 80 bit floating point value implementation for LDC compiler targetting MSVC
+// 80-bit floating point value implementation if the C/D compiler does not support them natively
 
 module dmd.root.longdouble;
 
-enum hasNativeFloat80 = real.sizeof > 8;
-
-static if (hasNativeFloat80)
-{
+static if (real.sizeof > 8)
     alias longdouble = real;
-}
+else
+    alias longdouble = longdouble_soft;
 
-static if (!hasNativeFloat80):
+// longdouble_soft needed when building the backend with
+// Visual C or the frontend with LDC on Windows
+version(CRuntime_Microsoft):
 extern(C++):
 nothrow:
+@nogc:
 
-align(2) struct longdouble
+version(D_InlineAsm_X86_64)
+    version = AsmX86;
+else version(D_InlineAsm_X86)
+    version = AsmX86;
+else
+    static assert(false, "longdouble_soft not supported on this platform");
+
+bool initFPU()
 {
-nothrow:
+    version(D_InlineAsm_X86_64)
+    {
+        // set precision to 64-bit mantissa and rounding control to nearest
+        asm nothrow @nogc pure
+        {
+            push    RAX;                 // add space on stack
+            fstcw   word ptr [RSP];
+            movzx   EAX,word ptr [RSP];  // also return old CW in EAX
+            and     EAX, ~0xF00;         // mask for PC and RC
+            or      EAX, 0x300;
+            mov     dword ptr [RSP],EAX;
+            fldcw   word ptr [RSP];
+            pop     RAX;
+        }
+    }
+    else version(D_InlineAsm_X86)
+    {
+        // set precision to 64-bit mantissa and rounding control to nearest
+        asm nothrow @nogc
```

Comment:
Not `pure` like the one above?

Reply:
`pure` is wrong above, the FPU control word is changed. I'll remove that.